consensus/src/sync/light/sync.rs

```diff
@@ -0,0 +1,187 @@
+
+#[derive(Clone)]
+/// This struct is used to request Epochs IDs (hashes) from other peers
+/// in order to determine their macro chain state relative to us
+pub(crate) struct EpochIds<T> {
+    /// Indicates if the latest epoch id that was queried was found in the peer's chain
+    pub locator_found: bool,
+    /// The most recent epoch ids (hashes)
+    pub ids: Vec<Blake2bHash>,
+    /// The most recent checkpoint block in the latest epoch (if any)
+    pub checkpoint: Option<Checkpoint>,
+    /// The first epoch number
+    pub first_epoch_number: usize,
+    /// The sender that created this struct
+    pub sender: T,
+}
+
+impl<T> EpochIds<T> {
+    #[inline]
+    pub(crate) fn checkpoint_epoch_number(&self) -> usize {
+        self.first_epoch_number + self.ids.len()
+    }
+
+    #[inline]
+    pub(crate) fn last_epoch_number(&self) -> usize {
+        self.checkpoint_epoch_number().saturating_sub(1)
+    }
+}
+
+/// This struct is used to track all the macro requests sent to a particular peer
+pub struct PeerMacroRequests {
+    /// Number of requests that have been fulfilled
+    completed_requests: usize,
+    /// A Queue used to track the requests that have been sent, and their respective result
+    queued_requests: VecDeque<(Blake2bHash, Option<Block>)>,
+}
+impl PeerMacroRequests {
+    pub fn new() -> Self {
+        Self {
+            completed_requests: 0,
+            queued_requests: VecDeque::new(),
+        }
+    }
+
+    // Pushes a new request into the queue
+    pub fn push_request(&mut self, block_hash: Blake2bHash) {
+        self.queued_requests.push_back((block_hash, None))
+    }
+
+    // Pops a request from the queue
+    pub fn pop_request(&mut self) -> Option<(Blake2bHash, Option<Block>)> {
+        self.queued_requests.pop_front()
+    }
+
+    // Returns true if the request was updated, false in case the request was not found
+    pub fn update_request(&mut self, block: Block) -> bool {
+        let position = self
+            .queued_requests
+            .iter()
+            .position(|(hash, _)| *hash == block.hash());
+
+        if let Some(position) = position {
+            if self.queued_requests[position].1.is_none() {
+                // A fulfilled request is only count once
+                self.completed_requests += 1;
+            }
+            // We update our block request.
+            // Note: If we receive a response more than once, we use the latest
+            self.queued_requests[position] = (block.hash(), Some(block));
+
+            true
+        } else {
+            false
+        }
+    }
+
+    // Returns true if all the requests have been completed
+    pub fn is_ready(&self) -> bool {
+        self.queued_requests.len() == self.completed_requests
+    }
+}
+
+/// The LightMacroSync is one type of MacroSync and it is essentially a stream,
+///  that operates on a per peer basis, emitting peers either as Outdated or Good.
```

Comment:
```suggestion
/// that operates on a per peer basis, emitting peers either as Outdated or Good.
```

Reply:
This was on purpose.. to have some sort of paragraphs